chain/chain/src/state_snapshot_actor.rs

```diff
@@ -0,0 +1,101 @@
+use actix::AsyncContext;
+use near_o11y::{handler_debug_span, OpenTelemetrySpanExt, WithSpanContext, WithSpanContextExt};
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardUId;
+use near_store::flat::FlatStorageManager;
+use near_store::ShardTries;
+use std::sync::Arc;
+
+/// Runs tasks related to state snapshots.
+pub struct StateSnapshotActor {
+    flat_storage_manager: FlatStorageManager,
+    tries: ShardTries,
+}
+
+impl StateSnapshotActor {
+    pub fn new(flat_storage_manager: FlatStorageManager, tries: ShardTries) -> Self {
+        Self { flat_storage_manager, tries }
+    }
+}
+
+impl actix::Actor for StateSnapshotActor {
+    type Context = actix::Context<Self>;
+}
+
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+struct MakeSnapshotRequest {
+    /// prev_hash of the last processed block.
+    prev_block_hash: CryptoHash,
+    /// Shards that need to be present in the snapshot.
+    shard_uids: Vec<ShardUId>,
+}
+
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+struct CompactSnapshotRequest {}
+
+/// Makes a state snapshot in the background.
+impl actix::Handler<WithSpanContext<MakeSnapshotRequest>> for StateSnapshotActor {
+    type Result = ();
+
+    fn handle(
+        &mut self,
+        msg: WithSpanContext<MakeSnapshotRequest>,
+        _ctx: &mut actix::Context<Self>,
+    ) -> Self::Result {
+        let (_span, msg) = handler_debug_span!(target: "state_snapshot", msg);
+        let MakeSnapshotRequest { prev_block_hash, shard_uids } = msg;
+
+        let run_compaction =
+            if let Err(err) = self.tries.make_state_snapshot(&prev_block_hash, &shard_uids) {
+                tracing::error!(target: "state_snapshot", ?err, "State snapshot creation failed");
+                false
+            } else {
+                true
+            };
+        assert!(
+            self.flat_storage_manager.set_flat_state_updates_mode(true),
+            "Failed to unlock flat state updates"
+        );
+        if run_compaction {
+            _ctx.address().do_send(CompactSnapshotRequest {}.with_span_context());
+        }
+    }
+}
+
+/// Runs compaction of the snapshot store.
+impl actix::Handler<WithSpanContext<CompactSnapshotRequest>> for StateSnapshotActor {
+    type Result = ();
+
+    fn handle(
+        &mut self,
+        msg: WithSpanContext<CompactSnapshotRequest>,
+        _ctx: &mut actix::Context<Self>,
+    ) -> Self::Result {
+        let (_span, _msg) = handler_debug_span!(target: "state_snapshot", msg);
+
+        if let Err(err) = self.tries.compact_state_snapshot() {
```

Comment:
I can't find this method, did you forget to add file?

Reply:
https://github.com/near/nearcore/pull/9090/files#diff-ba2b649b7a9490c5c6cc1adb9c8b34da9c92bbde788df9a6b9635d545f57235fR515